src/Middleware/RateLimiting/src/RateLimiterOptions.cs

```diff
@@ -44,5 +36,121 @@ public PartitionedRateLimiter<HttpContext> Limiter
+    public RateLimiterOptions AddPolicy<TPartitionKey>(string policyName, Func<HttpContext, RateLimitPartition<TPartitionKey>> partitioner)
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (partitioner == null)
+        {
+            throw new ArgumentNullException(nameof(partitioner));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        PolicyMap.Add(policyName, new DefaultRateLimiterPolicy(ConvertPartitioner<TPartitionKey>(partitioner), null));
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a new rate limiting policy with the given name.
+    /// </summary>
+    /// <param name="policyName">The name to be associated with the given TPolicy.</param>
+    public RateLimiterOptions AddPolicy<TPartitionKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPolicy>(string policyName) where TPolicy : IRateLimiterPolicy<TPartitionKey>
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        UnactivatedPolicyMap.Add(policyName, new PolicyTypeInfo { PolicyType = typeof(TPolicy), PartitionKeyType = typeof(TPartitionKey) });
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a new rate limiting policy with the given name.
+    /// </summary>
+    /// <param name="policyName">The name to be associated with the given <see cref="IRateLimiterPolicy{TPartitionKey}"/>.</param>
+    /// <param name="policy">The <see cref="IRateLimiterPolicy{TPartitionKey}"/> to be applied.</param>
+    public RateLimiterOptions AddPolicy<TPartitionKey>(string policyName, IRateLimiterPolicy<TPartitionKey> policy)
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        if (policy == null)
+        {
+            throw new ArgumentNullException(nameof(policy));
+        }
+
+        PolicyMap.Add(policyName, new DefaultRateLimiterPolicy(ConvertPartitioner<TPartitionKey>(policy.GetPartition), policy.OnRejected));
+
+        return this;
+    }
+
+    internal RateLimiterOptions InternalAddPolicy(string policyName, Func<HttpContext, RateLimitPartition<DefaultKeyType>> partitioner)
```

Comment:
Is there no way to just use the public version of AddPolicy?  I see that the RateLimiterOptionsExtensions are using this, but it seems better to have a single codepath, as it seems like the only difference is in what's passed to the DefaultRateLimiterPolicy between the public/internal?

Reply:
We always convert the users `TPartitionKey` to a `DefaultKeyType<TPartitionKey>` to avoid ambiguity - I don't think we want to provide users with a way to opt out of that. I guess we could just have one code path, which checks if `TPartitionKey` is `DefaultKeyType`, in which case we wouldn't do the conversion